aspnetcore-webapi/RateLimitingDemo/RateLimitingDemo.UsingCustomMiddleware/Middlewares/RateLimitingMiddleware.cs

```diff
@@ -0,0 +1,96 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
+using RateLimitingDemo.UsingCustomMiddleware.Decorators;
+using RateLimitingDemo.UsingCustomMiddleware.Extensions;
+using System.Net;
+
+namespace RateLimitingDemo.UsingCustomMiddleware.Middlewares;
+
+public class RateLimitingMiddleware
+{
+    private readonly RequestDelegate _next;
+
+    private readonly IDistributedCache _cache;
+    public RateLimitingMiddleware(RequestDelegate next, IDistributedCache cache)
+    {
+        _next = next;
+        _cache = cache;
+    }
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+
+
+        var rateLimitingDecorator = endpoint?.Metadata.GetMetadata<LimitRequests>();
+
+
+        if (rateLimitingDecorator is null)
+        {
+            await _next(context);
+            return;
+        }
+
+        var key = GenerateClientKey(context);
+
+        var clientStatistics = await GetClientStatisticsByKey(key);
+
+
+        if (clientStatistics != null)
+        {
+
+            if (DateTime.UtcNow < clientStatistics.LastSuccessfulResponseTime.AddSeconds(rateLimitingDecorator.TimeWindow) && clientStatistics.NumberOfRequestsCompletedSuccessfully == rateLimitingDecorator.MaxRequests)
```

Comment:
Merge these if statements or even better exit the method by checking if the clientStatistics is null first in order to avoid nested if statements.

Reply:
Merged the if conditions. Checking if clientStatistics is null would have required some duplicate code regarding updating the cache and calling the next delegate as it is the common step.